server/core/src/password_hash.cpp

```diff
@@ -0,0 +1,311 @@
+            std::vector<char> password_buf(_password.size() + 1, '\0');
+            const auto clear_password_buffer =
+                irods::at_scope_exit{[&password_buf] { std::ranges::fill(password_buf, '\0'); }};
+            std::strncpy(password_buf.data(), _password.c_str(), _password.size());
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+            *param++ = OSSL_PARAM_construct_octet_string(
+                OSSL_KDF_PARAM_PASSWORD, static_cast<void*>(password_buf.data()), _password.size());
+
+            // Construct salt parameter. The string must be copied to avoid const_cast in the octet string construction.
+            std::vector<char> salt_buf(_password.size() + 1, '\0');
+            std::strncpy(salt_buf.data(), _salt.c_str(), _salt.size());
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+            *param++ = OSSL_PARAM_construct_octet_string(
+                OSSL_KDF_PARAM_SALT, static_cast<void*>(salt_buf.data()), _salt.size());
+
+            // Complete construction of KDF parameters.
+            *param = OSSL_PARAM_construct_end();
+
+            // Merge the algorithm parameters and KDF parameters.
+            auto* merged_params = OSSL_PARAM_merge(params.data(), algorithm_params.data());
+            const auto free_merged_params = irods::at_scope_exit{[&merged_params] { OSSL_PARAM_free(merged_params); }};
+            if (nullptr == merged_params) {
+                THROW(SYS_INTERNAL_ERR, fmt::format("{}: Failed to merge OpenSSL parameters.", __func__));
+            }
+
+            log_server::trace("keylen={}, N={}, r={}, p={}", keylen, work_factor, resources, parallelism);
+
+            // Derive the key using the parameters constructed above. The result will be held in a buffer.
+            std::vector<unsigned char> hash_buf(keylen, 0);
+            if (EVP_KDF_derive(kctx, hash_buf.data(), keylen, merged_params) <= 0) {
+                THROW(CAT_PASSWORD_ENCODING_ERROR, fmt::format("{}: Failed to derive key for password.", __func__));
+            }
+
+            // Reset the object parameters so we can derive another key with this same instantiation, if needed.
+            reset();
+
+            // Return the hash buffer.
+            return hash_buf;
+        } // derive_key
+
+      private:
+        // KDF context for OpenSSL.
+        EVP_KDF_CTX* kctx;
+        // work/cost factor (N)
+        std::uint64_t work_factor;
+        // resources (r)
+        std::uint32_t resources;
+        // parallelism (p)
+        std::uint32_t parallelism;
+        // desired key length (dkLen)
+        std::uint32_t keylen;
+        // Parameters for scrypt algorithm. 4 elements needed for N, r, p, and "end" construct.
+        std::array<OSSL_PARAM, 4> algorithm_params;
+
+        inline auto free_kdf_context() -> void
+        {
+            if (nullptr != kctx) {
+                EVP_KDF_CTX_free(kctx);
+            }
```

Comment:
Probably good to set `kctx` to nullptr after freeing it.

Reply:
Done